### PR TITLE
Image status

### LINKED ISSUE #6: Report status to ImageList from each pod
Once pods finish, update status field of the ImageList CRD with timestamp, and operation details per node (status of image removal, message if error, node name, and image name).

### PR DESCRIPTION
**Pull request by ashnamehrotra**
- create ImageStatus objects in eraser pods
- update ImageJob status field with result of scheduled pods
- update ImageList status field with result of pod removals

### CODE REVIEW

File: controllers/imagestatus/imagestatus_controller.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
+/*
```
Comment (sozercan) [inline controllers/imagestatus/imagestatus_controller.go]: do we need this controller

File: config/samples/eraser_v1alpha1_imagejob.yaml [hunk 2]
```diff
@@ -11,3 +11,8 @@ spec:
           - name: eraser
             image: ashnam/remove-images:latest
           restartPolicy: OnFailure
+status:
```
Comment (sozercan) [inline config/samples/eraser_v1alpha1_imagejob.yaml]: we don't need these

File: pkg/eraser/go.mod [hunk 3]
```diff
@@ -3,11 +3,11 @@ module github.com/Azure/eraser/pkg/eraser
...
-	golang.org/x/net v0.0.0-20210428140749-89ef3d95e781 // indirect
-	golang.org/x/sys v0.0.0-20210603081109-ebe580a85c40 // indirect
 	github.com/Azure/eraser v0.0.0-20210720005525-9aab3f098186
 	google.golang.org/grpc v1.39.0
+	k8s.io/apimachinery v0.21.3
 	k8s.io/client-go v0.21.3
 	k8s.io/cri-api v0.21.3
 )
+
+replace github.com/Azure/eraser => github.com/ashnamehrotra/eraser v0.0.0-20210805193330-c5eec64189fc
```
Comment (sozercan) [inline pkg/eraser/go.mod]: TODO: remove before merging

**Comment by ashnamehrotra** [comment]
I think I need to make another PR to remove the replace after this PR gets merged since it won't compile because there are new status fields for imageJob

**Comment by cpuguy83** [comment]
@ashnamehrotra You can use relative paths for the replace rule.
e.g.

```
// assuming the current module is 2 levels deep
replace github.com/azure/eraser => ../../
```

**Comment by ashnamehrotra** [comment]
> @ashnamehrotra You can use relative paths for the replace rule.
> e.g.
> 
> ```
> // assuming the current module is 2 levels deep
> replace github.com/azure/eraser => ../../
> ```

after running go mod tidy when using relative paths it replaces it to the current one again

**Comment by sozercan** [comment]
that's a good idea to use relative path. i think we'll need to change `pkg/eraser/Dockerfile` and then copy the whole root then `cd` into `pkg/eraser` then build in there

File: controllers/imagejob/imagejob_controller.go [hunk 4]
```diff
@@ -189,3 +272,79 @@ func getMountPath(runtimeName string) string {
...
+	return true
+}
+
+func updateImageListStatus(nodeResult []eraserv1alpha1.NodeResult, imageList eraserv1alpha1.ImageList) {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
```
Comment (sozercan) [inline controllers/imagejob/imagejob_controller.go]: these 2 is duplicate in a few places, can we have a single one?

Comment (sozercan) [reply controllers/imagejob/imagejob_controller.go]: you can create one and then pass it around probably

File: controllers/imagejob/imagejob_controller.go [hunk 5]
```diff
@@ -189,3 +272,79 @@ func getMountPath(runtimeName string) string {
...
+	}
+
+	body, err := json.Marshal(imageList)
+	if err != nil {
+		log.Println(err)
+	}
+
+	// update imagelist object
+	_, err = clientset.RESTClient().Put().
+		AbsPath("apis/eraser.sh/v1alpha1").
```
Comment (sozercan) [inline controllers/imagejob/imagejob_controller.go]: these should use the consts

File: controllers/imagejob/imagejob_controller.go [hunk 6]
```diff
@@ -189,3 +272,79 @@ func getMountPath(runtimeName string) string {
...
+
+func updateImageListStatus(nodeResult []eraserv1alpha1.NodeResult, imageList eraserv1alpha1.ImageList) {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		log.Fatal(err)
```
Comment (sozercan) [inline controllers/imagejob/imagejob_controller.go]: for all these, it should be bubbled up (return err)

File: controllers/imagejob/imagejob_controller.go [hunk 7]
```diff
@@ -102,62 +109,138 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
...
+			}
+
+			updateJobStatus(*imageJob)
+
+			// transfer results from imageStatus objects to imageList
+			statusList := &eraserv1alpha1.ImageStatusList{}
+			err = r.List(ctx, statusList, &client.ListOptions{
+				Namespace: "eraser-system"})
+			if err != nil {
+				log.Println(err)
```
Comment (sozercan) [inline controllers/imagejob/imagejob_controller.go]: these should be returned (i have seen a few others like this), we can't proceed if this fails

File: controllers/imagejob/imagejob_controller.go [hunk 8]
```diff
@@ -189,3 +272,79 @@ func getMountPath(runtimeName string) string {
...
+	}
+
+	// update imagelist object
+	_, err = clientset.RESTClient().Put().
+		AbsPath("apis/eraser.sh/v1alpha1").
+		Namespace("eraser-system").
+		Name(imageList.Name).
+		Resource("imagelists").
+		SubResource("status").
+		Body(body).DoRaw(context.TODO())
```
Comment (sozercan) [inline controllers/imagejob/imagejob_controller.go]: pass in the context here instead of creating one each time

File: pkg/eraser/eraser.go [hunk 9]
```diff
@@ -143,6 +146,50 @@ func getImageClient(ctx context.Context, socketPath string) (pb.ImageServiceClie
 	return imageClient, conn, nil
 }
 
+func updateStatus(results []eraserv1alpha1.NodeCleanUpDetail) {
+	config, err := rest.InClusterConfig()
```
Comment (sozercan) [inline pkg/eraser/eraser.go]: same comment re: clientset

File: pkg/eraser/eraser.go [hunk 10]
```diff
@@ -184,49 +231,67 @@ func removeImages(c Client, socketPath string, targetImages []string) (err error
...
-		log.Println(img, "\t ", idMap[img])
-	}
 
 	nonRunningNames := make(map[string]struct{}, len(allImages)-len(runningImages))
 	remove := ""
 
 	for key := range nonRunningImages {
 		if idMap[key] != nil && len(idMap[key]) > 0 {
 			nonRunningNames[idMap[key][0]] = struct{}{}
+			// delete later, for testing
```
Comment (sozercan) [inline pkg/eraser/eraser.go]: TODO

File: pkg/eraser/eraser.go [hunk 11]
```diff
@@ -184,49 +231,67 @@ func removeImages(c Client, socketPath string, targetImages []string) (err error
...
 			remove = idMap[key][0]
 		}
 	}
 
-	// passing in a nonrunning image just for testing
+	// add an image to remove for testing
 	targetImages = append(targetImages, remove)
 
-	log.Println("\n\nTarget images:")
+	log.Println("\n\nTarget images: (1 additional added to ImageList to test remove)")
```
Comment (sozercan) [inline pkg/eraser/eraser.go]: TODO

File: pkg/eraser/eraser.go [hunk 12]
```diff
@@ -184,49 +231,67 @@ func removeImages(c Client, socketPath string, targetImages []string) (err error
...
-				return err
+				results = append(results, eraserv1alpha1.NodeCleanUpDetail{
+					ImageName: img,
+					Status:    "error",
+					Message:   err.Error(),
+				})
+			} else {
+				results = append(results, eraserv1alpha1.NodeCleanUpDetail{
+					ImageName: img,
+					Status:    "success",
```
Comment (sozercan) [inline pkg/eraser/eraser.go]: let's change these ("success", "error" etc) to types

File: pkg/eraser/eraser.go [hunk 13]
```diff
@@ -184,49 +231,67 @@ func removeImages(c Client, socketPath string, targetImages []string) (err error
...
+					ImageName: img,
+					Status:    "error",
+					Message:   "image not found",
+				})
 			}
 		}
-
 	}
 
+	updateStatus(results)
```
Comment (sozercan) [inline pkg/eraser/eraser.go]: handle errors returned

File: controllers/imagelist/imagelist_controller.go [hunk 14]
```diff
@@ -70,35 +71,45 @@ type Reconciler struct {
...
-						{
-							Name:            "remove-images",
-							Image:           "ashnam/remove_images:latest",
-							ImagePullPolicy: corev1.PullAlways,
-							Args:            []string{"--imagelist=" + req.Name},
+
+	imageList := &eraserv1alpha1.ImageList{}
+	err := r.Get(ctx, req.NamespacedName, imageList)
+	if err != nil {
+		log.Println(err)
```
Comment (sozercan) [inline controllers/imagelist/imagelist_controller.go]: nothing we can do, return here

File: pkg/eraser/eraser.go (hunk 12 above)
Comment (sozercan) [reply pkg/eraser/eraser.go]: similar to what you did with Job types

File: pkg/eraser/eraser.go [hunk 15]
```diff
@@ -143,6 +146,50 @@ func getImageClient(ctx context.Context, socketPath string) (pb.ImageServiceClie
...
+	_, err = clientset.RESTClient().Post().
+		AbsPath("apis/eraser.sh/v1alpha1").
+		Namespace("eraser-system").
+		Name(imageStatus.Name).
+		Resource("imagestatuses").
+		Body(body).DoRaw(context.TODO())
+
+	if err != nil {
+		log.Println(err)
+		log.Println("Could not create imagestatus for  node: ", os.Getenv("NODE_NAME"))
```
Comment (sozercan) [inline pkg/eraser/eraser.go:179]: return err

File: pkg/eraser/eraser.go [hunk 16]
```diff
@@ -143,6 +146,50 @@ func getImageClient(ctx context.Context, socketPath string) (pb.ImageServiceClie
...
+	}
+
+	body, err := json.Marshal(imageStatus)
+	if err != nil {
+		log.Println(err)
+	}
+
+	// create imageStatus object
+	_, err = clientset.RESTClient().Post().
+		AbsPath("apis/eraser.sh/v1alpha1").
```
Comment (sozercan) [inline pkg/eraser/eraser.go]: use consts

File: controllers/imagejob/imagejob_controller.go [hunk 17]
```diff
@@ -102,62 +109,138 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
...
+				})
+			}
+
+			imageList := &eraserv1alpha1.ImageListList{}
+			err = r.List(ctx, imageList)
+			if err != nil {
+				log.Println(err)
+			}
+
+			for _, l := range imageList.Items {
```
Comment (sozercan) [inline controllers/imagejob/imagejob_controller.go]: are we updating all imagelists here?

Comment (ashnamehrotra) [reply controllers/imagejob/imagejob_controller.go]: it should've only been the current one. Just changed it so imagejob receives imagelist name and uses that to get the list

File: controllers/imagejob/imagejob_controller.go [hunk 18]
```diff
@@ -102,62 +112,144 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
...
+			}
 		}
 
-		// TODO: check if pod fits and can be scheduled on node
-		err = r.Create(ctx, pod)
+	} else if imageJob.Status.Phase == eraserv1alpha1.PhaseRunning {
+		// get eraser pods
+		podList := &v1.PodList{}
+		err := r.List(ctx, podList, &client.ListOptions{
+			Namespace:     "eraser-system",
```
Comment (sozercan) [inline controllers/imagejob/imagejob_controller.go]: use the const here too

File: pkg/eraser/eraser.go [hunk 19]
```diff
@@ -143,7 +146,44 @@ func getImageClient(ctx context.Context, socketPath string) (pb.ImageServiceClie
...
+		return err
+	}
+
+	// create imageStatus object
+	_, err = clientset.RESTClient().Post().
+		AbsPath(apiPath).
+		Namespace(namespace).
+		Name(imageStatus.Name).
+		Resource("imagestatuses").
+		Body(body).DoRaw(context.TODO())
```
Comment (sozercan) [inline pkg/eraser/eraser.go]: pass the context here

File: controllers/imagejob/imagejob_controller.go [hunk 20]
```diff
@@ -102,62 +112,144 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
...
+				Failed:    failed,
+				Phase:     eraserv1alpha1.PhaseCompleted,
+			}
+
+			updateJobStatus(ctx, clientset, *imageJob)
+
+			// transfer results from imageStatus objects to imageList
+			statusList := &eraserv1alpha1.ImageStatusList{}
+			err = r.List(ctx, statusList, &client.ListOptions{
+				Namespace: "eraser-system"})
```
Comment (sozercan) [inline controllers/imagejob/imagejob_controller.go]: use the const here too

**Comment by ashnamehrotra** [comment]
I added a check to see if an image was running based on name input, since only ids were stored in the runningImages map. It uses a loop so that could be optimized later. I think that was causing some of the running images issues

File: api/v1alpha1/imagejob_types.go [hunk 21]
```diff
@@ -31,10 +31,29 @@ type ImageJobSpec struct {
 	ImageListName string             `json:"imageListName"`
 }
 
+type JobPhase string
+
+const (
```
Comment (cpuguy83) [inline api/v1alpha1/imagejob_types.go:37]: These will need a doc string.

File: api/v1alpha1/imagestatus_types.go [hunk 22]
```diff
@@ -17,10 +17,17 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+type RemovalStatus string
+
+const (
```
Comment (cpuguy83) [inline api/v1alpha1/imagestatus_types.go:23]: doc string

**Review by cpuguy83 — COMMENTED** [review]
Overall code wise this LGTM.

Tested this out and definitely it will help when these crd's are cluster scoped instead of namespaced.

The imagelist status I think is going to be problematically large since it has the status for each node+image to be cleaned up.
I wonder if we can bubble up a summary into the imagelist status (maybe a copy of what's in imagejob) instead, with a link to the imagejob, and in imagejob link to each imagestatuses.

**Review by sozercan — APPROVED** [review]
LGTM

**Comment by sozercan** [comment]
> The imagelist status I think is going to be problematically large since it has the status for each node+image to be cleaned up.
I wonder if we can bubble up a summary into the imagelist status (maybe a copy of what's in imagejob) instead, with a link to the imagejob, and in imagejob link to each imagestatuses.

@cpuguy83 can you open an issue for this please?

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Pull request merged**